gramps/gen/filters/optimizer.py

```diff
@@ -17,121 +18,93 @@
-    to include or exclude.
-    """
 
-    def __init__(self, filter):
-        """
-        Initialize the collection of selected_handles in the filter list.
-        """
-        self.all_selected_handles = []
-        self.walk_filters(filter, False, self.all_selected_handles)
+def intersection(sets: List[Set[PrimaryObjectHandle]]) -> Set[PrimaryObjectHandle]:
+    # sort the sets by length, shortest first.
+    # with intersection, the shortest of the starting sets determines the
+    # maximum size of the result
+    sorted_sets = sorted(sets, key=lambda s: len(s))
+    return set.intersection(*sorted_sets)
 
-    def walk_filters(self, filter, parent_invert, result):
-        """
-        Recursively walk all of the filters/rules and get
-        rules with selected_handles
-        """
-        current_invert = parent_invert if not filter.invert else not parent_invert
-        LOG.debug(
-            "walking, filter: %s, invert=%s, parent_invert=%s",
-            filter,
-            filter.invert,
-            parent_invert,
-        )
-        rules_with_selected_handles = []
-        for item in filter.flist:
-            if hasattr(item, "find_filter"):
-                rule_filter = item.find_filter()
-                if rule_filter is not None:
-                    self.walk_filters(
-                        rule_filter,
-                        current_invert,
-                        result,
-                    )
-            elif hasattr(item, "selected_handles"):
-                rules_with_selected_handles.append(set(item.selected_handles))
-        if rules_with_selected_handles:
-            LOG.debug(
-                "filter %s: parent_invert=%s, invert=%s, op=%s, number of rules with selected_handles=%s",
-                filter,
-                parent_invert,
-                filter.invert,
-                filter.logical_op,
-                len(rules_with_selected_handles),
-            )
-            result.append(
-                (
-                    current_invert,
-                    filter.logical_op,
-                    rules_with_selected_handles,
-                )
-            )
 
-    def get_handles(self):
-        """
-        Returns handles_in, and handles_out.
+def union(sets: List[Set[PrimaryObjectHandle]]) -> Set[PrimaryObjectHandle]:
+    return set.union(*sets)
 
-        `handles_in` is either None, or a set of handles to include.
-        if it is None, then there is no evidence to only include
-        particular handles. If it is a set, then those in the set
-        are a superset of the items that will match.
 
-        `handles_out` is a set. If any handle is in the set, it will
-        not be included in the final results.
+class Optimizer:
+    def compute_potential_handles_for_filter(
+        self, filter: GenericFilter
+    ) -> Tuple[
+        Union[Set[PrimaryObjectHandle], None], Union[Set[PrimaryObjectHandle], None]
```

Comment:
```suggestion
        Set[PrimaryObjectHandle] | None, Set[PrimaryObjectHandle] | None
```

In other places, `|` is used in preference to the more verbose `Union[]`